rnnalyse/activations/activation_reader.py

```diff
@@ -22,34 +23,166 @@ class ActivationReader:
     Attributes
     ----------
     activations_dir : str
-    labels : np.ndarray
-        Numpy array containing the extracted labels
-    data_len : int
-        Number of extracted activations
+    label_path : str
+    activations : Optional[np.ndarray]
+        Numpy array of activations that are currently read into ram.
+    _labels : Optional[np.ndarray]
```

Comment:
I thought that these were the underscores that @Kaleidophon was referring to before, but apparently I misunderstood. Why are these attributes underscored? Is it a convention to denote they are private?

Reply:
Oh wait, I think I misunderstood @Kaleidophon 's remark as well. It is indeed a convention to denote variables that are intended for internal use, this blog explains it well: https://dbader.org/blog/meaning-of-underscores-in-python

I use it there because `self.labels` now refers to the `@property` which returns the actual array that is stored in `._labels`.

However, in `LanguageModel` I mainly used it because putting an `_` in front of it stopped my checker from complaining about singleton variables. 

Thinking about this, it makes much more sense of course to use `**kwargs` in the `LanguageModel` init, I'll update it.

Reply:
Sorry for all the rambling, but the only thing I'm uncertain about right now is that `LanguageModel` doesn't tell you at all which arguments it's expecting from its subclasses. Maybe `*args` isn't too perfect after all, putting the explicit arguments there did make that more clear.

What do you think?

Reply:
Adding `@abstractmethod` to the init seems to stop pylint from complaining, and also makes the expected arguments more explicit. I think this is the way to go then.

Reply:
I think if you're using `@property` it is fine to use something like `_labels`. In the docstring I would just refer to `labels` because the `labels` function decorated with `@property`  essentially functions like an attribute.

Reply:
And why was Pylint complaining about init? Using `@abstractmethod` on init doesn't seem to make a lot of sense to me

Reply:
I guess it did because a non-abstractmethod is expected to do something with its arguments. I got the idea of making init an abstractmethod from this SO discussion: https://stackoverflow.com/questions/44800659/python-abstract-class-init

So to couple it back to why we're doing all this: we want the `LanguageModel` to be an abstract class which ensures its subclasses adhere to the same init structure (model/vocab/module/device). To ensure this we don't want to use generic `*args` in the init, as this would allow a subclass to input any kind of stuff it wants there.

Furthermore, we don't want to simply set these arguments as class attributes in `LanguageModel`, as for instance the `model` argument is only the path to the model and it is up to the subclass to load in the model itself. On the other hand we could of course add an abstractmethod to `LanguageModel` that should take care of loading in the model/vocab/module stuff. This might make the usecase a bit more explicit.

Reply:
Okay I just fiddled around a bit and I have to say that I was wrong and just putting `ABC` as a superclass doesn't prevent you from instantiating it, you have to indeed add an `@abstractmethod` to init. However, if you do override the (default super-)constructor you can basically pass an abritrary signature and it still works (your IDE might warn you though, I guess that is what we're aiming for?).

The example I is [here](https://trinket.io/python3/7fd5b76925).

Reply:
Yeah nice, it starts to make sense now.

If you'd override the init constructor with a different signature pylint will complain, so that's nice. It is already complaining right now about the `forward` overriding the `forward` of `nn.module`, which might be kinda related to the point you raised about training a new language model and all that. Food for thought, I can imagine we want to change it again somewhere in the future